### PR TITLE
Relocate compiler flag for hip version into HipBuild()

### PR DESCRIPTION
**Pull request by asroy**
Relocate compiler flag for hip version into HipBuild(), so that it's not part of the key for kernel binary cache.

Also refactors some code.

----
This is to resolve the problem mentioned at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/499#issuecomment-707391991.

### CODE REVIEW

**Comment by atamazov** [comment]
Description updated.

**Review by atamazov — COMMENTED** [review]
Please note that there is another method. The compiler version can be passed in HipBuild(). This would not affect the binary cache key, i.e. works as if the compiler version macro is predefined by the compiler. This approach looks more attractive to me:
- The version, once supplied, could be used for various source code adjustments in the future.
- The binary cache is not affected at all. For example, HIP kernels built with >= 3.10.20405 will be found in the cache even if user's compiler is < 3.10. Right now this PR doesn't allow this.

**Comment by asroy** [comment]
@atamazov I moved the hip version flag into HipBuild()

**Review by atamazov — APPROVED** [review]
Thanks @asroy

**Comment by daniellowell** [comment]
@asroy @atamazov 
Macro not defined in the file:
```/var/jenkins/workspace/ibs_MIOpen_remove_ck_hip_version/src/comgr.cpp:704:74: error: use of undeclared identifier 'HIP_PACKAGE_VERSION_FLAT'

                       (" -DHIP_PACKAGE_VERSION_FLAT=") + std::to_string(HIP_PACKAGE_VERSION_FLAT);

                                                                         ^

/var/jenkins/workspace/ibs_MIOpen_remove_ck_hip_version/src/comgr.cpp:716:74: error: use of undeclared identifier 'HIP_PACKAGE_VERSION_FLAT'

                       (" -DHIP_PACKAGE_VERSION_FLAT=") + std::to_string(HIP_PACKAGE_VERSION_FLAT);

                                                                         ^

2 errors generated when compiling for gfx900.
```
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/remove_ck_hip_version/2/pipeline

**Review by atamazov — APPROVED** [review]

**Comment by daniellowell** [comment]
@asroy This PR has a merge conflict now.

**Comment by atamazov** [comment]
Conflict is due to my commit. Resolved.

**Review by atamazov — APPROVED** [review]

**Comment by asroy** [comment]
@daniellowell This PR passed CI

File: src/hip/hip_build_utils.cpp [hunk 1]
```diff
@@ -335,4 +339,9 @@ bool external_tool_version_t::operator>=(const external_tool_version_t& rhs) con
         return false;
 }
 
+bool external_tool_version_t::operator<(const external_tool_version_t& rhs) const
+{
+    return rhs >= *this;
```
Comment (atamazov) [inline src/hip/hip_build_utils.cpp:344]: :bug: Bug here -- should be `>`. I'll fix it in my next PR.

**Pull request merged**